services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt

```diff
@@ -113,38 +167,287 @@ class HakemusServiceITest : DatabaseTest() {
+                    USERID
+                )
+            }
+        }
+
+        @Test
+        fun `throws exception when the application has been sent to Allu`() {
+            val application = createHakemus { alluid = 21 }
 
-    private fun Assert<CustomerWithContactsResponse>.isCompanyCustomerWithOneContact(
-        orderer: Boolean
-    ) {
-        prop(CustomerWithContactsResponse::customer)
-            .prop(CustomerResponse::type)
-            .isEqualTo(CustomerType.COMPANY)
-
-        prop(CustomerWithContactsResponse::contacts)
-            .single()
-            .prop(ContactResponse::orderer)
-            .isEqualTo(orderer)
+            val exception =
+                assertThrows<ApplicationAlreadySentException> {
+                    hakemusService.updateHakemus(
+                        application.id!!,
+                        HakemusUpdateRequestFactory.createFilledJohtoselvityshakemusUpdateRequest(),
+                        USERID
+                    )
+                }
+
+            assertThat(exception.message)
+                .isEqualTo("Application is already sent to Allu, id=${application.id}, alluid=21")
+        }
+
+        @Test
+        fun `does not create a new audit log entry when the application has not changed`() {
+            val application = createHakemus()
+            val originalAuditLogSize = auditLogRepository.findByType(ObjectType.APPLICATION).size
+
+            hakemusService.updateHakemus(
+                application.id!!,
+                HakemusUpdateRequestFactory
+                    .createJohtoselvityshakemusUpdateRequestFromApplicationEntity(application),
+                USERID
+            )
+
+            val applicationLogs = auditLogRepository.findByType(ObjectType.APPLICATION)
+            assertThat(applicationLogs).hasSize(originalAuditLogSize)
+        }
+
+        @Test
+        fun `throws exception when there are invalid geometry in areas`() {
+            val application = createHakemus()
+
+            val exception =
+                assertThrows<ApplicationGeometryException> {
+                    hakemusService.updateHakemus(
+                        application.id!!,
+                        HakemusUpdateRequestFactory
+                            .createJohtoselvityshakemusUpdateRequestFromApplicationEntity(
+                                application
+                            )
+                            .copy(areas = listOf(intersectingArea)),
+                        USERID
+                    )
+                }
+
+            Assertions.assertEquals(
+                """Invalid geometry received when updating application id=${application.id},
+                | reason = Self-intersection,
+                | location = {"type":"Point","coordinates":[25494009.65639264,6679886.142116806]}"""
+                    .trimMargin()
+                    .replace("\n", ""),
+                exception.message
+            )
```

Comment:
Here as well it would be nice to stick with assertk.

Check the exception message character-by-character might not be the most productive approach, since it will break easily. Checking that the message contains the important details like the application ID, reason and maybe the coordinates is less fragile and easier to read.

Reply:
You're right. This was also due to copy-pasting from ApplicationServiceITest. Will change.